### ISSUE TITLE
Che workspace takes too much time to load

### ISSUE THREAD

**Issue by deanone** [eclipse-researchlabs/smartclide-ide-front-end#16]
Dear @goncalorolo,

I have made several tests on the platform and I noticed that even for existing services (i.e., for services that the Che workspace has already been created) the workspace takes too much time to load. It is about 2 to 3 minutes.

So, I think that this is something that we need to check since it affects the whole user experience.

**Comment by goncalorolo** [comment]
Even though it is not as severe as for new workspaces, starting a workspace usually takes some time, which is something we have had to put up with since the first experiments in the Che dashboard.

It is true that it undermines the user experience but there seems to be no simple solution for this.

**Comment by deanone** [comment]
I'll investigate this more thoroughly. You know, yesterday I was presenting the platform to the ORW and during the live demo the Che workspace for a test project that I had already created before the meeting took like 3-4 minutes to load. This is way too long. Any developer that would experience such a delay on a development tool, most probably he will quickly drop it and move to the next.

**Comment by goncalorolo** [comment]
> I'll investigate this more thoroughly. You know, yesterday I was presenting the platform to the ORW and during the live demo the Che workspace for a test project that I had already created before the meeting took like 3-4 minutes to load. This is way too long. Any developer that would experience such a delay on a development tool, most probably he will quickly drop it and move to the next.

Yes, that is the usual behaviour. The only way to avoid showing that in the presentation would have been to open the workspace beforehand and leave it open (you can leave Che-Theia in the IDE frontend by clicking the logo of the project in the header because that changes the page but leaves the workspace running, whereas the "Close" option in the sidebar stops it).

**Comment by deanone** [comment]
> > I'll investigate this more thoroughly. You know, yesterday I was presenting the platform to the ORW and during the live demo the Che workspace for a test project that I had already created before the meeting took like 3-4 minutes to load. This is way too long. Any developer that would experience such a delay on a development tool, most probably he will quickly drop it and move to the next.
> 
> Yes, that is the usual behaviour. The only way to avoid showing that in the presentation would have been to open the workspace beforehand and leave it open (you can leave Che-Theia in the IDE frontend by clicking the logo of the project in the header because that changes the page but leaves the workspace running, whereas the "Close" option in the sidebar stops it).

Yes, but the think is that I wanted to actually show the workspace launching. That's why I created a service beforehand because I thought that the Che workspace will launch (more or less) quickly, if it was not the first time launching the workspace.

I believe that a workaround exists. We just need to find it.

**Comment by goncalorolo** [comment]
> Yes, but the think is that I wanted to actually show the workspace launching. That's why I created a service beforehand because I thought that the Che workspace will launch (more or less) quickly, if it was not the first time launching the workspace.

Oh, I see. Well, yes, the first time you run the workspace is when it takes the longest. The following attempts are a bit better but still noticeably slow.

**Comment by deanone** [comment]
Dear @goncalorolo ,

We really need to have a closer look at this issue. Today I tried to load the workspaces of existing services that I have created through the platform, and each load took to much time. For example, in one case, I had to wait for more than 10 minutes!!

Any ideas on what may cause this issue?

**Comment by goncalorolo** [comment]
In this particular case, I think workspaces were taking too much time to start due to a Che downtime (which also led to #32). I have just tried starting an existing workspace and creating a new one and starting it afterwards and both processes didn't take more than 2 minutes each.

**Comment by deanone** [comment]
@goncalorolo indeed I just tried to load an existing workspace and I took around 2.5 minutes.

However, I want to ask, is this normal in general? I mean, has it been reported somewhere that Che workspaces are very slow to load?

**Comment by deanone** [comment]
@goncalorolo any thoughts on that?

**Comment by goncalorolo** [comment]
> * `Optimizing workspace plug-ins`: not sure if @goncalorolo can do anything about that?

The devfile templates we currently provide only use the minimum subset of plugins needed. Therefore, we cannot optimise this any further.

> * `Reducing the number of public endpoints`: not sure if @goncalorolo can do anything about that?

Similarly to the previous bullet point, there are no unused public endpoints being exposed.

**Comment by deanone** [comment]
> @goncalorolo @deanone Regarding the image puller: the [documentation](https://www.eclipse.org/che/docs/stable/administration-guide/defining-the-list-of-images-to-pull/) states:
> 
> > For faster workspace startup times, consider pulling workspace related images such as universal-developer-image, che-code`, and che-gateway.
> 
> When looking at https://plugin-registry-smartclide-che.che.smartclide.eu/v3/external_images.txt none of the recommended images appear there.
> 
> Would it make sense to try to cache any of the other images from that list?

@philipreimer @goncalorolo I think it is worth trying to pull the recommended images universal-developer-image, che-code`, and che-gateway
